### PR TITLE
Implement Swift serialization and deserialization of Clang types

### PR DESCRIPTION
**Pull request by rjmccall**
As part of this, we have to change the type export rules to prevent `@convention(c)` function types from being used in exported interfaces if they aren't serializable.  This is a more conservative version of the original rule I had, which was to import such function-pointer types as opaque pointers.  That rule would've completely prevented importing function-pointer types defined in bridging headers and so simply doesn't work, so we're left trying to catch the unsupportable cases retroactively.  This has the unfortunate consequence that we can't necessarily serialize the internal state of the compiler, but that was already true due to normal type uses of aggregate types from bridging headers; if we can teach the compiler to reliably serialize such types, we should be able to use the same mechanisms for function types.
    
This PR doesn't flip the switch to use Clang function types by default, so many of the clang-function-type-serialization FIXMEs are still in place.

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please test.

File: test/ClangImporter/unserializable-clang-function-types.swift [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// RUN: %target-swift-frontend -typecheck -swift-version 5 -emit-module-interface-path - -sdk %clang-importer-sdk -enable-library-evolution %s -experimental-print-full-convention -verify
+
+import ctypes
+
+public var f1 : UnserializableFunctionPointer?
+// expected-error@-1 {{cannot export the underlying C type of the function type 'UnserializableFunctionPointer' (aka '@convention(c) () -> Optional<OpaquePointer>'); it may use anonymous types or types defined outside of a module}}
```
Comment (varungandhi-apple) [inline test/ClangImporter/unserializable-clang-function-types.swift]: Is this theoretically a source break, because people might have code that is exposing opaque/non-modular types publicly in Swift? If yes, do you think this is unlikely to happen in practice?

Comment (rjmccall) [reply test/ClangImporter/unserializable-clang-function-types.swift]: It is theoretically a source break, but I wouldn't expect it to be a problem in practice as long as we catch the major cases of unimportable types.

**Comment by varungandhi-apple** [comment]
Could you elaborate on this with an example:

> This has the unfortunate consequence that we can't necessarily serialize the internal state of the compiler, but that was already true due to normal type uses of aggregate types from bridging headers; if we can teach the compiler to reliably serialize such types, we should be able to use the same mechanisms for function types.

Are bridging headers special because they are not part of a module (I think?)?

Also, it feels like this knowledge of exceptional cases should be documented somewhere in the code itself, but I'm not entirely sure where.

File: include/swift/AST/PrettyStackTrace.h [hunk 2]
```diff
@@ -132,6 +136,17 @@ class PrettyStackTraceType : public llvm::PrettyStackTraceEntry {
   virtual void print(llvm::raw_ostream &OS) const;
 };
 
+/// PrettyStackTraceClangType - Observe that we are processing a
+/// specific Clang type.
+class PrettyStackTraceClangType : public llvm::PrettyStackTraceEntry {
```
Comment (varungandhi-apple) [inline include/swift/AST/PrettyStackTrace.h]: Thanks for adding this, it'll probably come in handy for debugging.

File: lib/AST/PrettyStackTrace.cpp [hunk 3]
```diff
@@ -209,6 +210,15 @@ void swift::printTypeDescription(llvm::raw_ostream &out, Type type,
   if (addNewline) out << '\n';
 }
 
+void PrettyStackTraceClangType::print(llvm::raw_ostream &out) const {
+  out << "While " << Action << ' ';
+  if (TheType == nullptr) {
+    out << "NULL clang type!\n";
+    return;
+  }
+  TheType->dump(out);
```
Comment (varungandhi-apple) [inline lib/AST/PrettyStackTrace.cpp]: This incurs a non-header-only dependency on the Clang AST/Type.h header in our AST. The "solution" I'd used earlier was to pass in an ASTContext and grab a ClangModuleLoader off that. I'm going to submit the PR today or tomorrow to re-land the Clang types in SIL, so if this is getting merged after that, you can use the same helper method.

File: lib/AST/ClangTypeConverter.cpp [hunk 4]
```diff
@@ -726,3 +730,17 @@ clang::QualType ClangTypeConverter::convert(Type type) {
   Cache.insert({type, result});
   return result;
 }
+
+void ClangTypeConverter::registerExportedClangDecl(Decl *swiftDecl,
+                                             const clang::Decl *clangDecl) {
+  assert(clangDecl->getCanonicalDecl() == clangDecl);
```
Comment (varungandhi-apple) [inline lib/AST/ClangTypeConverter.cpp]: Could you add a string with `&&` documenting why this is expected to hold?

File: lib/Serialization/Serialization.cpp [hunk 5]
```diff
@@ -4221,6 +4251,90 @@ void Serializer::writeASTBlockEntity(Type ty) {
...
+    if (value) {
+      id = S.addDeclBaseNameRef(
+               S.getASTContext().getIdentifier(value->getName()));
+    }
+    Record.push_back(id);
+  }
+
+  void writeStmtRef(const clang::Stmt *stmt) {
+    // Always read null; the checker should prevent this from being
+    // called with non-null statements.
```
Comment (varungandhi-apple) [inline lib/Serialization/Serialization.cpp]: Could you add an if null check + `llvm::report_fatal_error` here instead of having an empty body?

File: include/swift/AST/ClangModuleLoader.h [hunk 6]
```diff
@@ -111,6 +155,22 @@ class ClangModuleLoader : public ModuleLoader {
...
+  /// Follow a stable serialization path.
+  virtual const clang::Decl *
+  lookupStableSerializationPath(const StableSerializationPath &path) const = 0;
+
+  /// Determine whether the given type is serializable.
+  ///
+  /// If \c checkCanonical is true, checks the canonical type,
+  /// not the given possibly-sugared type.
+  virtual bool isSerializable(const clang::Type *type,
+                              bool checkCanonical) const = 0;
```
Comment (varungandhi-apple) [inline include/swift/AST/ClangModuleLoader.h]: What do you think about removing the boolean flag and having the caller compute the canonical type if it needs to? Right now, there's only 1 caller which calls it with the second argument set to true. It feels like it is trying to do multiple things which are not really its responsibility. Instead, we could have a comment on `ClangImporter::isSerializable` giving examples for when the two serializable status (for a canonical type vs a sugared type) might be different, leaving the responsibility to the caller.

**Comment by varungandhi-apple** [comment]
Can you add a couple of test cases with `-use-clang-function-types`?

1. serialization + llvm-bcanalyzer
2. serialization + deserialization

[I don't think that requires SIL work to land, does it?]

File: include/swift/AST/ClangModuleLoader.h [hunk 7]
```diff
@@ -40,6 +43,47 @@ enum class ClangTypeKind {
...
+
+      /// A C++ namespace
+      Namespace
+    };
+
+    ModuleDecl *M;
+    SmallVector<std::pair<IdentifierNamespace, Identifier>, 2> Path;
+  };
+private:
+  TaggedUnion<void, const Decl *, ExternalPath> Union;
```
Comment (varungandhi-apple) [inline include/swift/AST/ClangModuleLoader.h]: Is there a specific reason why you decided to add a null state here instead of having the different functions return an `Optional<StableSerializationPath>`? This option is less verbose?

File: lib/Serialization/Deserialization.cpp [hunk 8]
```diff
@@ -5174,6 +5182,17 @@ class TypeDeserializer {
     if (!diffKind.hasValue())
       MF.fatal();
 
+    const clang::FunctionType *clangFunctionType = nullptr;
+    if (clangFunctionTypeID) {
+      auto clangType = MF.getClangType(clangFunctionTypeID);
+      if (!clangType)
+        return clangType.takeError();
+      clangFunctionType =
+        dyn_cast_or_null<clang::FunctionType>(clangType.get());
```
Comment (varungandhi-apple) [inline lib/Serialization/Deserialization.cpp]: This seems insufficient in the presence of block pointers. Is that something we are not going to do in the future? [I think we discussed this and the answer was no, but double-checking in case I am misremembering. If we are not doing that, I should update a bunch of code/comments.]

**Comment by rjmccall** [comment]
> Are bridging headers special because they are not part of a module (I think?)?

Right, in two different ways.  They're not part of a module, so we have no way of figuring out where they're from; and that's supposed to be okay because you aren't supposed to use declarations from bridging headers in module interfaces, which is currently enforced by simply not allowing modules to have bridging headers except for the final executable (which then doesn't have an interface).

The general restriction should be enforced in the type-checker, and we want to do that as part of  the implementation-only imports checking, but we haven't yet.  I suspect it will look a lot like this check.

File: lib/AST/PrettyStackTrace.cpp (hunk 3 above)
Comment (rjmccall) [reply lib/AST/PrettyStackTrace.cpp]: Per offline discussion, this is okay.

File: include/swift/AST/ClangModuleLoader.h (hunk 7 above)
Comment (rjmccall) [reply include/swift/AST/ClangModuleLoader.h]: Yeah, I found it simpler.

File: lib/Serialization/Deserialization.cpp (hunk 8 above)
Comment (rjmccall) [reply lib/Serialization/Deserialization.cpp]: We should absolutely still do block pointers.  I didn't think about tthem when I was writing this because the existing `clangFunctionType` variable was a `const clang::FunctionType *`, apparently because `SILFunctionType::ExtInfo` requires that.  So I actually can't generalize this given the current interfaces; since you have an outstanding patch in that area, I'll let you handle it.

File: include/swift/AST/ClangModuleLoader.h (hunk 6 above)
Comment (rjmccall) [reply include/swift/AST/ClangModuleLoader.h]: I actually wanted to encourage the caller to think about this.

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
> Can you add a couple of test cases with `-use-clang-function-types`?
> 
> 1. serialization + llvm-bcanalyzer
> 2. serialization + deserialization
> 
> [I don't think that requires SIL work to land, does it?]

I'll flesh out the serialization tests some more tomorrow.  It shouldn't need SIL work.

File: test/Inputs/clang-importer-sdk/usr/include/ctypes.h [hunk 9]
```diff
@@ -224,6 +224,17 @@ typedef struct Dummy {
 
 Dummy * (*getFunctionPointer3(void))(Dummy *);
 
+// These two function types should be serializable despite the struct
+// declarations be incomplete and therefore (currently) unimiportable.
```
Comment (varungandhi-apple) [inline test/Inputs/clang-importer-sdk/usr/include/ctypes.h]: Typo: "be" -> "being"

File: lib/Serialization/Serialization.cpp [hunk 10]
```diff
@@ -4809,6 +4929,7 @@ void Serializer::writeAST(ModuleOrSourceFile DC,
     writeOffsets(Offsets, SubstitutionMapsToSerialize);
     writeOffsets(Offsets, NormalConformancesToSerialize);
     writeOffsets(Offsets, SILLayoutsToSerialize);
+    writeOffsets(Offsets, ClangTypesToSerialize);
```
Comment (varungandhi-apple) [inline lib/Serialization/Serialization.cpp]: Could you move this line to earlier, right after `writeOffsets(Offsets, TypesToSerialize);`? That way the order matches the order for `wroteSomething |= blah`.

File: lib/Serialization/Serialization.cpp [hunk 11]
```diff
@@ -4032,11 +4057,15 @@ class Serializer::TypeSerializer : public TypeVisitor<TypeSerializer> {
   void visitFunctionType(const FunctionType *fnTy) {
     using namespace decls_block;
 
+    auto resultType = S.addTypeRef(fnTy->getResult());
+    auto clangType = S.addClangTypeRef(fnTy->getClangFunctionType());
+
     // FIXME: [clang-function-type-serialization] Serialize the clang type here
```
Comment (varungandhi-apple) [inline lib/Serialization/Serialization.cpp]: We can remove this `FIXME` now. 😃

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test source compatibility

File: lib/Serialization/Serialization.cpp (hunk 10 above)
Comment (rjmccall) [reply lib/Serialization/Serialization.cpp]: Sure

File: test/Inputs/clang-importer-sdk/usr/include/ctypes.h (hunk 9 above)
Comment (rjmccall) [reply test/Inputs/clang-importer-sdk/usr/include/ctypes.h]: Thanks, fixed this and the other on this line.

File: include/swift/ClangImporter/SwiftAbstractBasicReader.h [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    SmallVector<clang::IdentifierInfo *, 4> chunks;
+    for (unsigned i = 0, e = std::max(numArgs, 1U); i != e; ++i)
+      chunks.push_back(asImpl().readIdentifier());
+
+    return getASTContext().Selectors.getSelector(numArgs, chunks.data());
+  }
+
+  clang::SourceLocation readSourceLocation() {
+    // Always read null.
+    return clang::SourceLocation();
```
Comment (varungandhi-apple) [inline include/swift/ClangImporter/SwiftAbstractBasicReader.h]: Should this have a `TODO: Also track source locations. That requires Foo Bar.` or are we planning to keep it this way? I remember one of my patches needed a source location to print the Clang diagnostic and we just made up a fake one or something...

File: test/Serialization/clang-function-types.swift [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+// CHECK-LABEL: sil hidden @$s4main5test2yyF
+func test2() {
+  use_fp_internally()
+}
+// CHECK-LABEL: } // end sil function '$s4main5test2yyF'
+
+// CHECK-LABEL: sil public_external [canonical] @$s24def_clang_function_types17use_fp_internallyyyF
+// CHECK:         enum $Optional<@convention(c) () -> Optional<OpaquePointer>>, #Optional.none!enumelt
+// CHECK:         [[FN:%.*]] = function_ref @$s24def_clang_function_types9use_inout3argyxz_tlF : $@convention(thin) <τ_0_0> (@inout τ_0_0) -> ()
+// FIXME: something is very wrong in the printer here
```
Comment (varungandhi-apple) [inline test/Serialization/clang-function-types.swift]: Is this something that is temporarily present for while you're working on the tests/making CI pass? What is going wrong here?

Comment (rjmccall) [reply test/Serialization/clang-function-types.swift]: I didn't change anything about the printer, so I don't know what's going on.

Comment (varungandhi-apple) [reply test/Serialization/clang-function-types.swift]: I guess my question is: what is the symptom you are seeing? I couldn't understand that from the test case text. I understand you don't have a diagnosis/solution because you didn't fiddle with the printer.

File: include/swift/ClangImporter/SwiftAbstractBasicReader.h (hunk 12 above)
Comment (rjmccall) [reply include/swift/ClangImporter/SwiftAbstractBasicReader.h]: I don't think there will ever be a meaningful way for us to stably serialize a Clang source location.  Fortunately, they do not generally appear in Clang types, especially in prototypes.

Th serialization/deserialization logic is designed to have a precondition of passing the isSerializable check.  It looks like I'm missing a comment to that effect on the deserialization side.

Comment (varungandhi-apple) [reply include/swift/ClangImporter/SwiftAbstractBasicReader.h]: Gotcha'. I think your additional context about 

> I don't think there will ever be a meaningful way for us to stably serialize a Clang source location. Fortunately, they do not generally appear in Clang types, especially in prototypes.

would be useful to have in a comment next to the code, perhaps here itself.

File: include/swift/AST/ClangModuleLoader.h [hunk 14]
```diff
@@ -40,6 +43,66 @@ enum class ClangTypeKind {
   ObjCProtocol,
 };
 
+/// A path for serializing a declaration.
+class StableSerializationPath {
+public:
+  struct ExternalPath {
+    enum StepKind {
```
Comment (varungandhi-apple) [inline include/swift/AST/ClangModuleLoader.h]: What do you think about renaming this `ComponentKind`? That's the typical terminology I've seen around filepaths.

File: include/swift/AST/ClangModuleLoader.h [hunk 15]
```diff
@@ -40,6 +43,66 @@ enum class ClangTypeKind {
...
+      /// A struct/union/class
+      Struct,
```
Comment (varungandhi-apple) [inline include/swift/AST/ClangModuleLoader.h]: What do you think about renaming this to `Record` to match the Clang name? I see that most of the other ones (all of them?) match the names for the corresponding Clang decls.

Comment (rjmccall) [reply include/swift/AST/ClangModuleLoader.h]: Good idea.

File: include/swift/AST/ClangModuleLoader.h [hunk 16]
```diff
@@ -40,6 +43,66 @@ enum class ClangTypeKind {
...
+  const Decl *getSwiftDecl() const {
+    assert(isSwiftDecl());
+    return Union.get<const Decl*>();
+  }
+
+  bool isExternalPath() const { return Union.isa<ExternalPath>(); }
+  const ExternalPath &getExternalPath() const {
+    assert(isExternalPath());
+    return Union.get<ExternalPath>();
+  }
```
Comment (varungandhi-apple) [inline include/swift/AST/ClangModuleLoader.h]: Thoughts on folding these 4 methods into only 2 which return `Optional`s?

File: include/swift/ClangImporter/ClangImporter.h [hunk 17]
```diff
@@ -422,6 +422,16 @@ class ClangImporter final : public ClangModuleLoader {
...
+
+  StableSerializationPath
+  findStableSerializationPath(const clang::Decl *decl) const override;
+
+  const clang::Decl *
+  resolveStableSerializationPath(
+                            const StableSerializationPath &path) const override;
+
+  bool isSerializable(const clang::Type *type,
+                      bool checkCanonicial) const override;
```
Comment (varungandhi-apple) [inline include/swift/ClangImporter/ClangImporter.h]: Typo: "checkCanonicial" -> "checkCanonical"..

File: include/swift/ClangImporter/SwiftAbstractBasicReader.h [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+// Clang's AbstractBasicReader interface, paralleling the behavior defined
+// in SwiftAbstractBasicWriter.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_CLANGIMPORTER_SWIFTABSTRACTBASICREADER_H
+#define SWIFT_CLANGIMPORTER_SWIFTABSTRACTBASICREADER_H
+
+#include "clang/AST/AbstractTypeReader.h"
+// AbstractBasicReader.h doesn't include what it uses
```
Comment (varungandhi-apple) [inline include/swift/ClangImporter/SwiftAbstractBasicReader.h]: It is unclear whether this line applies to the header below or above, could you make it an end-of-line-comment instead?

File: include/swift/ClangImporter/SwiftAbstractBasicWriter.h [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  void writeBool(bool value) {
+    asImpl().writeUInt64(uint64_t(value));
+  }
+
+  void writeUInt32(uint32_t value) {
+    asImpl().writeUInt64(uint64_t(value));
+  }
+
+  void writeSelector(clang::Selector selector) {
+    if (selector.isNull()) {
```
Comment (varungandhi-apple) [inline include/swift/ClangImporter/SwiftAbstractBasicWriter.h]: [I thought selectors were a value level thing, not a type-level thing.] Do Clang types potentially have selectors in them? If yes, when would we encounter a null selector? [Maybe we should have a small comment here with 1-2 examples...]

File: include/swift/AST/ClangModuleLoader.h (hunk 16 above)
Comment (rjmccall) [reply include/swift/AST/ClangModuleLoader.h]: I find this pattern of accessors that assert you're working with the right kind of thing generally work better in C++.  Also, `ExternalPath` is a non-trivial type.

File: include/swift/ClangImporter/SwiftAbstractBasicReader.h (hunk 18 above)
Comment (rjmccall) [reply include/swift/ClangImporter/SwiftAbstractBasicReader.h]: Expanded the comment to make it clearer.

File: include/swift/ClangImporter/SwiftAbstractBasicWriter.h (hunk 19 above)
Comment (rjmccall) [reply include/swift/ClangImporter/SwiftAbstractBasicWriter.h]: It can't happen with types directly, but it's necessary for the underlying serialization abstractions because selectors can appear in common-currency types like `DeclarationName`.  By handling it, we're prepared if we start serializing expressions, which can appear in types.

File: test/Serialization/clang-function-types.swift (hunk 13 above)
Comment (rjmccall) [reply test/Serialization/clang-function-types.swift]: Long-standing bug, apparently.

File: include/swift/ClangImporter/SwiftAbstractBasicWriter.h (hunk 19 above)
Comment (rjmccall) [reply include/swift/ClangImporter/SwiftAbstractBasicWriter.h]: It can't happen with types directly, but it's necessary for the underlying serialization abstractions because selectors can appear in common-currency types like DeclarationName. By handling it, we're prepared if we start serializing expressions, which can appear in types.

**Comment by rjmccall** [comment]
@swift-ci Please test.

File: include/swift/AST/ClangModuleLoader.h (hunk 16 above)
Comment (varungandhi-apple) [reply include/swift/AST/ClangModuleLoader.h]: I don't understand how the fact that `ExternalPath` is a non-trivial type is relevant here... what does that have to do with whether we choose to return an `Optional` or not?

**Comment by rjmccall** [comment]
@swift-ci Please test OS X

File: lib/AST/ClangTypeConverter.cpp [hunk 20]
```diff
@@ -726,3 +730,19 @@ clang::QualType ClangTypeConverter::convert(Type type) {
   Cache.insert({type, result});
   return result;
 }
+
+void ClangTypeConverter::registerExportedClangDecl(Decl *swiftDecl,
+                                             const clang::Decl *clangDecl) {
+  assert(clangDecl->getCanonicalDecl() == clangDecl &&
```
Comment (varungandhi-apple) [inline lib/AST/ClangTypeConverter.cpp]: This check can be replaced with `clangDecl->isCanonicalDecl()`.

File: include/swift/AST/ClangModuleLoader.h [hunk 21]
```diff
@@ -40,6 +43,66 @@ enum class ClangTypeKind {
   ObjCProtocol,
 };
 
+/// A path for serializing a declaration.
+class StableSerializationPath {
+public:
+  struct ExternalPath {
+    enum ComponentKind {
```
Comment (varungandhi-apple) [inline include/swift/AST/ClangModuleLoader.h]: Could you add `print`/`dump` methods for these three types? (Or say only for `StableSerializationPath` and `ExternalPath`)

File: test/Serialization/clang-function-types.swift (hunk 13 above)
Comment (rjmccall) [reply test/Serialization/clang-function-types.swift]: The bug was that it was printing `Optional<@convention(c) () -> Optional<SomePointer>>` as `@convention(c) () -> SomePointer??`.

File: include/swift/AST/ClangModuleLoader.h (hunk 16 above)
Comment (rjmccall) [reply include/swift/AST/ClangModuleLoader.h]: Correct.  It would be better to not have an API that can only be implemented by copying non-trivial values out of the container.

File: lib/Serialization/Serialization.cpp [hunk 22]
```diff
@@ -585,6 +590,25 @@ serialization::TypeID Serializer::addTypeRef(Type ty) {
...
+  if (ty) {
+    // Try to serialize the non-canonical type, but fall back to the
+    // canonical type if necessary.
+    auto loader = getASTContext().getClangModuleLoader();
+    if (!ty->isCanonicalUnqualified()) {
+      if (!loader->isSerializable(ty, /*check canonical*/ false)) {
+        ty = ty->getCanonicalTypeInternal().getTypePtr();
+      }
+    }
+    if (!loader->isSerializable(ty, /*check canonical*/ false)) {
```
Comment (varungandhi-apple) [inline lib/Serialization/Serialization.cpp]: The logic here seems a bit hard to follow. For example, this looks like if we have a non-canonical type, then we will do a second check for `isSerializable` even if the first check passed, which is probably not intended. Instead, you could refactor it to something like:

```
bool isSerializable;
if (ty->isCanonicalUnqualified()) {
  isSerializable = loader->isSerializable(ty, /*checkCanonical*/ false);
} else if (loader->isSerializable(ty, /*checkCanonical*/ false)) {
  isSerializable = true;
} else {
  ty = ty->getCanonicalTypeInternal().getTypePtr();
  isSerializable = loader->isSerializable(ty, /*checkCanonical*/ false);
}
if (!isSerializable) {
  // report fatal error
}
```

This might seem a bit too pedestrian (e.g. I could've gotten rid of the `isSerializable = true;` by initializing it with `true` at the beginning instead and having a `!loader->isSerializable` to compress it further) but the benefit IMO is that there is only 1 negated condition at the end, all the others are positive checks so it is easy to see that the value is initialized in all the branches.

**Comment by varungandhi-apple** [comment]
I've mostly looked at everything except `Serializability.cpp` -- I'll go over that tomorrow or later today.

**Comment by rjmccall** [comment]
@swift-ci Please smoke test

File: lib/Serialization/Serialization.cpp (hunk 22 above)
Comment (rjmccall) [reply lib/Serialization/Serialization.cpp]: Reworked

**Comment by rjmccall** [comment]
@swift-ci Please smoke test OS X

**Comment by rjmccall** [comment]
> I've mostly looked at everything except `Serializability.cpp` -- I'll go over that tomorrow or later today.

I'm going to merge, but if you have further suggestions, I'd be happy to apply them in a follow-up.
